fs: Handle absolute pathname for --stdin-filename

garrmcnu opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

What is the purpose of this change? What does it change?
--------------------------------------------------------

<!--
Describe the changes here, as detailed as needed.
-->
Return valid directory info from Lstat() for parent directories of the specified filename. Previously only "/" and "." were valid directories.
Also set directory mode as this is checked by archiver.


Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

<!--
Link issues and relevant forum posts here.
-->
Closes #2063
Closes #1809

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

garrmcnu committed:
fs: Update directory check in reader tests (#2063)

fd0 commented:
Just checking in (since you haven't checked the last box): is this PR ready for review? Are you done? Or do you need help with anything?

garrmcnu commented:
Thanks for the check-in, yes this is ready for review. If this approach looks good, I believe it does handle absolute pathnames. The only question I had was should relative directory paths be allowed for `--stdin-filename`?
With this fix, a relative directory will still fail because the snapshot path is always pre-pended with "/" and so will always be absolute. But I think supplying a relative path in this context doesn't make sense, so giving an error for a relative directory would seem correct. But good to get other thoughts on that?

rawtaz commented:
Why not just expand the relative path name so the corresponding absolute one, and use that? Then it's supported and is up to the user to decide how they want to specify the path, and it's in line with how you specify paths in other parts of restic.

fd0 commented:
> Why not just expand the relative path name so the corresponding absolute one, and use that?

I don't think that's a good idea, then the path in the snapshot depends on the current directory, which may be surprising. I'd just take the path as an absolute path, and be done with it:
 *  `/foo/bar` becomes `/foo/bar`
 *  `foo/bar` becomes `/foo/bar`
 *  `./foo` becomes `/foo`
 *  `../../foo/bar` becomes `/foo/bar`

And so on. That's (IMHO) the least surprising thing we can do.

fd0 commented:
Whoopsi, that was the wrong button :)

fd0 commented on internal/fs/fs_reader_test.go:
```diff
@@ -317,6 +327,49 @@ func TestFSReader(t *testing.T) {
 		})
 	}
+
+	var dirTests = []struct {
```
I'd split this out into its own test function, maybe `TestFSReaderDir`?

garrmcnu replied:
yes, that's clearer.

fd0 commented on internal/fs/fs_reader_test.go:
```diff
@@ -151,8 +152,8 @@ func verifyDirectoryContentsFI(t testing.TB, fs FS, dir string, want []os.FileIn
 func checkFileInfo(t testing.TB, fi os.FileInfo, filename string, modtime time.Time, mode os.FileMode, isdir bool) {
-	if fi.IsDir() {
-		t.Errorf("IsDir returned true, want false")
+	if fi.IsDir() != isdir {
```
Oh, good catch!

fd0 commented on internal/fs/fs_reader_test.go:
```diff
@@ -151,8 +152,8 @@ func verifyDirectoryContentsFI(t testing.TB, fs FS, dir string, want []os.FileIn
-	if fi.IsDir() {
-		t.Errorf("IsDir returned true, want false")
+	if fi.IsDir() != isdir {
+		t.Errorf("IsDir returned %t, want %t", fi.IsDir(), isdir)
```
We need another test, `os.FileInfo` defines that `Name()` returns the base of the file name, so `path.Base(filename)`:
```suggestion
		t.Errorf("IsDir returned %t, want %t", fi.IsDir(), isdir)
	}
	
	if path.Base(fi.Name()) != fi.Name() {
		t.Errorf("Name() returned is not base: want %q, got %q", path.Base(fi.Name()), fi.Name())
```

fd0 replied:
We also need to check `fi.Name() != path.Base(filename)` below (unfortunately I can't add a suggestion for it).

garrmcnu replied:
Ah yes, I completely missed that, thanks!

fd0 commented on internal/fs/fs_reader.go:
```diff
@@ -103,17 +103,32 @@ func (fs *Reader) Stat(name string) (os.FileInfo, error) {
-	case "/", ".":
+	getDirInfo := func(name string) os.FileInfo {
 		fi := fakeFileInfo{
 			name:    name,
```
This needs to be `path.Base(name)`:
```suggestion
			name:    path.Base(name),
```

fd0 reviewed (commented):
I've added a few comments what we need to change, in general it's a great contribution! In order to handle the relative path issue, I'd suggest passing the path given as `--stdin-filename` through `path.Clean(path.Join("/", opts.StdinFilename))` in `cmd_backup.go:526`. This will do the right thing.

Thanks!

EDIT: I pressed the wrong button (again), this was meant to be a change request.

garrmcnu committed:
fs: Add file info base name check in reader tests (#2063)

garrmcnu committed:
fs: Handle absolute pathname for --stdin-filename

Return valid directory info from Lstat() for parent directories of the
specified filename. Previously only "/" and "." were valid directories.

Also set directory mode as this is checked by archiver.

Closes #2063

garrmcnu committed:
backup: Convert relative pathname for --stdin-filename to absolute (#2063)

garrmcnu commented:
Thanks for the feedback, that sound good, will make those changes.

On the relative path, yes treating it as an absolute path looks like a clean solution and that works, thanks.
One minor point on the suggested change, I believe path.Join() already calls Clean() so don't think it needs to be called on the joined path.

garrmcnu commented:
Thanks for all the help on this.

And of course, thanks for restic, great backup solution!
